federation/pkg/federation-controller/ingress/ingress-controller.go

```diff
@@ -0,0 +1,301 @@
+	ic.ingressFederatedInformer = util.NewFederatedInformer(
+		client,
+		func(cluster *federation_api.Cluster, targetClient federation_release_1_4.Interface) (cache.Store, framework.ControllerInterface) {
+			return framework.NewInformer(
+				&cache.ListWatch{
+					ListFunc: func(options api.ListOptions) (pkg_runtime.Object, error) {
+						return targetClient.Extensions().Ingresses("<all-namespaces>").List(options)
+					},
+					WatchFunc: func(options api.ListOptions) (watch.Interface, error) {
+						return targetClient.Extensions().Ingresses("<all-namespaces>").Watch(options)
+					},
+				},
+				&extensions.Ingress{},
+				controller.NoResyncPeriodFunc(),
+				// Trigger reconcilation whenever something in federated cluster is changed. In most cases it
+				// would be just confirmation that some ingress opration suceeded.
+				util.NewTriggerOnMetaAndSpecChangesPreproc(
+					func(obj pkg_runtime.Object) { ic.deliverIngressObj(obj, IngressReviewDelay, 0) },
+					func(obj pkg_runtime.Object) { util.SetClusterName(obj, cluster.Name) },
+				))
+		},
+
+		&util.ClusterLifecycleHandlerFuncs{
+			ClusterAvailable: func(cluster *federation_api.Cluster) {
+				// When new cluster becomes available process all the ingresss again.
+				ic.clusterDeliverer.DeliverAt(allClustersKey, nil, time.Now().Add(ClusterAvailableDelay))
+			},
+		},
+	)
+
+	// Federated updeater along with Create/Update/Delete operations.
+	ic.federatedUpdater = util.NewFederatedUpdater(ic.ingressFederatedInformer,
+		func(client federation_release_1_4.Interface, obj pkg_runtime.Object) error {
+			ingress := obj.(*extensions_v1beta1.Ingress)
+			_, err := client.Extensions().Ingresses("<all-namespaces>").Create(ingress)
+			return err
+		},
+		func(client federation_release_1_4.Interface, obj pkg_runtime.Object) error {
+			ingress := obj.(*extensions_v1beta1.Ingress)
+			_, err := client.Extensions().Ingresses("<all-namespaces>").Update(ingress)
+			return err
+		},
+		func(client federation_release_1_4.Interface, obj pkg_runtime.Object) error {
+			ingress := obj.(*extensions_v1beta1.Ingress)
+			err := client.Extensions().Ingresses("<all-namespaces>").Delete(ingress.Name, &api.DeleteOptions{})
+			return err
+		})
+	return ic
+}
+
+func (ic *IngressController) Start() {
+	ic.ingressInformerController.Run(ic.stopChan)
+	ic.ingressFederatedInformer.Start()
+	ic.ingressDeliverer.StartWithHandler(func(item *util.DelayingDelivererItem) {
+		ni := item.Value.(*ingressItem)
+		ic.reconcileIngress(ni.ingress, ni.trial)
+	})
+	ic.clusterDeliverer.StartWithHandler(func(_ *util.DelayingDelivererItem) {
+		ic.reconcileIngresssOnClusterChange()
+	})
+}
+
+func (ic *IngressController) Stop() {
+	ic.ingressFederatedInformer.Stop()
+	close(ic.stopChan)
+}
+
+func (ic *IngressController) deliverIngressObj(obj interface{}, delay time.Duration, trial int64) {
+	ingress := obj.(*extensions.Ingress)
+	ic.deliverIngress(ingress.Name, delay, trial)
```

Comment:
Agreed.  Will do.   The controller that this is modelled on is for namespaces themselves, which are not scoped by namespace, hence this.